nemo/collections/asr/parts/submodules/tdt_beam_decoding.py

```diff
@@ -0,0 +1,793 @@
+                # it will produce two hypotheses with the same token sequence but different scores.
+                kept_hyps = self.merge_duplicate_hypotheses(kept_hyps)
+
+                if len(hyps) > 0:
+                    # Keep those hypothesis that have scores greater than next search generation
+                    hyps_max = float(max(hyps, key=lambda x: x.score).score)
+                    kept_most_prob = sorted(
+                        [hyp for hyp in kept_hyps if hyp.score > hyps_max],
+                        key=lambda x: x.score,
+                    )
+                    # If enough hypotheses have scores greater than next search generation,
+                    # stop beam search.
+                    if len(kept_most_prob) >= beam:
+                        kept_hyps = kept_most_prob
+                        break
+                else:
+                    # If there are no hypotheses in a current frame, keep only `beam` best hypotheses for the next search generation.
+                    kept_hyps = sorted(kept_hyps, key=lambda x: x.score, reverse=True)[:beam]
+        return self.sort_nbest(kept_hyps)
+
+    def modified_adaptive_expansion_search(
+        self,
+        encoder_outputs: torch.Tensor,
+        encoded_lengths: torch.Tensor,
+        partial_hypotheses: Optional[Hypothesis] = None,
+    ) -> List[Hypothesis]:
+        """
+        Modified Adaptive Exoansion Search algorithm for TDT models.
+        Based on/modified from https://ieeexplore.ieee.org/document/9250505.
+        Supports N-gram language model shallow fusion.
+
+        Args:
+            encoder_outputs: encoder outputs (batch, features, timesteps).
+            encoded_lengths: lengths of the encoder outputs.
+            partial_hypotheses: partial hypotheses.
+
+        Returns:
+            nbest_hyps: N-best decoding results
+        """
+        if partial_hypotheses is not None:
+            raise NotImplementedError("Support for `partial_hypotheses` is not implemented.")
+
+        beam = min(self.beam_size, self.vocab_size)
+        beam_state = self.decoder.initialize_state(
+            torch.zeros(1, device=encoder_outputs.device, dtype=encoder_outputs.dtype)
+        )  # [L, B, H], [L, B, H] for LSTMS
+
+        # Initialize first hypothesis for the beam (blank).
+        start_hyp = Hypothesis(
+            y_sequence=[self.blank],
+            score=0.0,
+            dec_state=self.decoder.batch_select_state(beam_state, 0),
+            timestep=[-1],
+            length=0,
+            last_frame=0,
+        )
+        init_tokens = [start_hyp]
+
+        # Cache decoder results to avoid duplicate computations.
+        cache = {}
+
+        # Decode a batch of beam states and scores
+        beam_decoder_output, beam_state = self.decoder.batch_score_hypothesis(init_tokens, cache)
+        state = beam_state[0]
+
+        # Setup ngram LM:
+        if self.ngram_lm:
```

Comment:
I think it we merge this block with the next `if self.ngram_lm`, it would be more readable